sdk/dotnet/Pulumi.Automation/ProjectRuntimeOptions.cs

```diff
@@ -29,5 +32,34 @@ public class ProjectRuntimeOptions
         /// A string that specifies the path to a virtual environment to use when running the program.
         /// </summary>
         public string? VirtualEnv { get; set; }
+
+        internal static IEqualityComparer<ProjectRuntimeOptions> Comparer { get; } = new ProjectRuntimeOptionsComparer();
+
+        private sealed class ProjectRuntimeOptionsComparer : IEqualityComparer<ProjectRuntimeOptions>
+        {
+            bool IEqualityComparer<ProjectRuntimeOptions>.Equals(ProjectRuntimeOptions? x, ProjectRuntimeOptions? y)
+            {
+                if (x == null)
+                {
+                    return y == null;
+                }
+
+                if (y == null)
+                {
+                    return x == null;
+                }
+
+                return x.TypeScript == y.TypeScript && x.Binary == y.Binary && x.VirtualEnv == y.VirtualEnv;
+            }
+
+            int IEqualityComparer<ProjectRuntimeOptions>.GetHashCode(ProjectRuntimeOptions obj)
+            {
+                var hash = new HashCode();
+                hash.Add(obj.TypeScript);
+                hash.Add(obj.Binary);
+                hash.Add(obj.VirtualEnv);
+                return hash.ToHashCode();
```

Comment:
Can consider using [HashCode.Combine](https://docs.microsoft.com/en-us/dotnet/api/system.hashcode.combine?view=net-5.0#System_HashCode_Combine__3___0___1___2_)

```suggestion
                return HashCode.Combine(obj.TypeScript, obj.Binary, obj.VirtualEnv);
```